### PR TITLE
2.1.1 - Featured Article Image Cropper Meta Box

### PR DESCRIPTION
**Pull request by ronalfy**
I'm working on an image cropper that can work for featured article images on a project. I had some JS for one of my project ages ago that actually got cropping to work. It took a bit of cleanup and bug fixes, but it works really well.

Screen Shot 2022-02-02 at 3 12 48 PM

I've also coded it so it can be used more than once for any post type. So if the theme has a different slider or needs a specific image size and ratio, you can use it wherever there's a post type. The code could also help others. The cropping mechanism is muy complicado.

Draft PR because I am not finished.

### CODE REVIEW

File: functions.php [hunk 1]
```diff
@@ -50,6 +50,10 @@ function course_maker_localization_setup() {
 // Add the custom meta boxes.
 require_once CHILD_THEME_DIR . '/lib/metaboxes.php';
 
+// Add the the attachment meta box class.
+require_once CHILD_THEME_DIR . '/lib/meta-attachments.php';
+require_once CHILD_THEME_DIR . '/lib/meta-attachments-init.php';
+
```
Comment (ronalfy) [inline functions.php:56]: I have the main class file, which contains the featured article files. One is a class, and the other an inisitalizer.

File: js/featured-article-image.js [hunk 2]
```diff
@@ -0,0 +1,182 @@
+// any global scripts needed for theme goes here
+;
+(function($) {
+
+    'use strict';
+	var __ = wp.i18n.__;
+	var cropOptions = ( attachment, controller ) => {
+		var control = controller.get( 'control' );
+
+		console.log( attachment );
```
Comment (ronalfy) [inline js/featured-article-image.js:10]: I need to remove these :)

File: js/featured-article-image.js [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+		console.log( control );
+
```
Comment (ronalfy) [inline js/featured-article-image.js:12]: Crazy math ahead. In hindsight, I could have named them better.

File: js/featured-article-image.js [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+				xImg = realWidth;
+			}
+			xInit = xImg;
+			yInit = xInit / ratio;
+		}
+
+		var x1 = ( realWidth - xInit ) / 2;
+		var y1 = ( realHeight - yInit ) / 2;
+
+		var imgSelectOptions = {
```
Comment (ronalfy) [inline js/featured-article-image.js:46]: This is the bulk of crop init.

File: js/featured-article-image.js [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+			keys: true,
+			instance: true,
+			persistent: true,
+			imageWidth: realWidth,
+			imageHeight: realHeight,
+			x1,
+			y1,
+			x2: xInit + x1,
+			y2: yInit + y1,
+			aspectRatio: '40:21',
```
Comment (ronalfy) [inline js/featured-article-image.js:57]: The aspect ratio is important here. Cropping won't work without it.

File: js/featured-article-image.js [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+		};
+		return imgSelectOptions;
+	};
+
+	/**
+	 * Set the attachment image.
+	 * @param {object} image Image object after cropping.
+	 */
+	function setAttachmentImage( image ) {
+		console.log( image );
```
Comment (ronalfy) [inline js/featured-article-image.js:67]: Sorry :(

File: js/featured-article-image.js [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+	}
+
+	var cropControl = {
```
Comment (ronalfy) [inline js/featured-article-image.js:70]: This is where params are set. Flex width and flex height allow the image to maintain a crop if smaller or larger.

Width/Height is the desired width/height for the image.

File: js/featured-article-image.js [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+		// If the destination width is equal to or greater than the cropped image width
+		// then the user does not need to crop the image...
+		if ( imgW <= dstW ) {
+			return false;
+		}
+
+		return true;
+	};
+
+	function handleUpload( e ) {
```
Comment (ronalfy) [inline js/featured-article-image.js:115]: This is what happens when a featured button gets pressed. Lots going on here.

File: js/featured-article-image.js [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+					suggestedHeight: 630,
+				} ),
+				new wp.media.controller.CustomizeImageCropper( {
+					imgSelectOptions: cropOptions,
+					control: cropControl,
+				} ),
+			],
+		} );
+
+		mediaUploader.on( 'cropped', function( croppedImage ) {
```
Comment (ronalfy) [inline js/featured-article-image.js:139]: Event fires when image is cropped.

File: js/featured-article-image.js [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+			setAttachmentImage( {
+				id: croppedImage.attachment_id,
+				url: croppedImage.url,
+			} );
+		} );
+
+		mediaUploader.on( 'insert', function() {
```
Comment (ronalfy) [inline js/featured-article-image.js:146]: Callback for the Selection of the image in the media library.

File: js/featured-article-image.js [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+			var json = mediaUploader.state().get( 'selection' ).first().toJSON();
+			var url = json.sizes.hasOwnProperty( 'thumbnail' )
+				? json.sizes.thumbnail.url
+				: json.sizes.full.url;
+				setAttachmentImage( {
+				id: json.id,
+				url,
+			} );
+		} );
+		mediaUploader.on( 'select', function() {
```
Comment (ronalfy) [inline js/featured-article-image.js:156]: When the image in the media uploader is selected.

File: js/featured-article-image.js [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+					setAttachmentImage( {
+					id: json.id,
+					url,
+				} );
+				mediaUploader.close();
+			} else {
+				mediaUploader.setState( 'cropper' );
+			}
+		} );
+		mediaUploader.open();
```
Comment (ronalfy) [inline js/featured-article-image.js:177]: Opens the media dialogue.

File: lib/meta-attachments-init.php [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+ * @package Course Maker Pro
+ * @author  thebrandiD
+ * @license GPL-2.0+
+ * @link    https://thebrandidthemes.com/
+ */
+
+/**
+ * Initialize the meta box attachment class.
+ */
+function course_maker_add_attachment_image_init() {
```
Comment (ronalfy) [inline lib/meta-attachments-init.php:14]: This initializes the meta boxes.

File: lib/meta-attachments.php [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+  * Class for handling meta attachments.
+  */
+class Meta_Attachments {
+
+	/**
+	 * Unique ID for this meta box.
+	 *
+	 * @var string $id
+	 */
+	private string $id = '';
```
Comment (ronalfy) [inline lib/meta-attachments.php:21]: private vars. We'll need them later.

File: lib/meta-attachments.php [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+	 *     An array of arguments.
+	 *
+	 *     @type string $meta_key Meta key to register (stored in post meta)
+	 *     @type array $post_types Array of post types to list this meta box on.
+	 *     @type string $meta_label Label for insert attachment button (see add_meta_box).
+	 *     @type string $meta_priority Priority of meta box (see add_meta_box).
+	 *     @type string $meta_location Location of meta box (see add_meta_box).
+	 * }
+	 */
+	public function __construct( $args ) {
```
Comment (ronalfy) [inline lib/meta-attachments.php:87]: Class is instantiated and sets class vars. Has defaults, which can be overridden.

File: lib/meta-attachments.php [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+		// Set class variables.
+		foreach ( $args as $key => $value ) {
+			$this->{ $key } = $value;
+		}
+	}
+
+	/**
+	 * Run the hooks associated with this class.
+	 */
+	public function run_hooks() {
```
Comment (ronalfy) [inline lib/meta-attachments.php:113]: This adds the actions for the metaboxes with callback. 

Finally we load the media scripts needed.

File: lib/meta-attachments.php [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+		add_action( 'admin_enqueue_scripts', 'course_maker_display_featured_article_image_scripts' );
+	}
+
+	/**
+	 * Registers the Attachment Meta box.
+	 *
+	 * @param object $post The current post object.
+	 */
+	public function add_meta_box_callback( $post ) {
+		add_meta_box(
```
Comment (ronalfy) [inline lib/meta-attachments.php:128]: Dynamic meta box creation.

File: lib/meta-attachments.php [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+			$this->meta_priority
+		);
+	}
+
+	/**
+	 * Displays the meta box.
+	 *
+	 * @param object $post The current post object.
+	 */
+	public function display_meta_box( $post ) {
```
Comment (ronalfy) [inline lib/meta-attachments.php:143]: Meta box display. Has a button that when pressed, opens the media dialogue so a user can select an image.

File: lib/meta-attachments.php [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+
+		</div>
+		<?php
+	}
+}
+
+/**
+ * Load post admin scripts for the featured article image.
+ */
+function course_maker_display_featured_article_image_scripts( $hook ) {
```
Comment (ronalfy) [inline lib/meta-attachments.php:170]: This hasn't been placed in the class yet. I need to complete this part.

Comment (ronalfy) [reply lib/meta-attachments.php:170]: Just need to create a jQuery plugin so the script can handle multiple inputs (e.g., 2 meta boxes on the same post type).

**Review by saragreenlaw — APPROVED** [review]

**Review by saragreenlaw — APPROVED** [review]

**Pull request merged**